MaxText/inference/page_manager.py

```diff
@@ -1,232 +1,582 @@
+
+  Args:
+    num_pages: The total number of available pages in the global pool.
+    max_page_groups: The maximum number of page groups (concurrent sequences/requests)
+      the system can track.
+    max_pages_per_group: The maximum number of pages that can be allocated to
+      a single page group (determines the size of the second dimension of `page_map`).
+
+  Returns:
+    An initialized `PageState` object with all values set to their defaults (zeros/False).
+  """
+  initial_page_status = jnp.zeros((num_pages,), dtype=jnp.int32)
+  initial_page_status = initial_page_status.at[0].set(1)  # Workaround page 0
   return PageState(
-      page_status=jnp.zeros((num_pages,), dtype=jnp.int32),
-      page_map=jnp.zeros((num_slots, max_pages_per_slot), dtype=jnp.int32),
-      sequence_lengths=jnp.zeros((num_slots,), dtype=jnp.int32),
-      num_pages_used=jnp.zeros((num_slots,), dtype=jnp.int32),
-      current_page=jnp.zeros((num_slots,), dtype=jnp.int32),
-      current_page_position=jnp.zeros((num_slots,), dtype=jnp.int32),
+      page_status=initial_page_status,
+      page_map=jnp.zeros((max_page_groups, max_pages_per_group), dtype=jnp.int32),
+      num_pages_used=jnp.zeros((max_page_groups,), dtype=jnp.int32),
+      sequence_lengths=jnp.zeros((max_page_groups,), dtype=jnp.int32),
+      active_page=jnp.zeros((max_page_groups,), dtype=jnp.int32),
+      has_active_page=jnp.zeros((max_page_groups,), dtype=jnp.bool_),
+      active_page_position=jnp.zeros((max_page_groups,), dtype=jnp.int32),
   )
 
 
-class PageManager:
-  """Manages paged attention mechanism for efficient sequence processing.
+def _find_next_free_page_index(page_status: PagesInt1d) -> ScalarInt:
+  """Finds the index of the next available free page in the global pool.
 
-  The PageManager implements a virtual memory-like system for attention, where the
-  attention context is divided into fixed-size pages. This allows efficient handling
-  of variable-length sequences and helps manage memory usage during inference.
+  Searches the `page_status` array for the first occurrence of 0 (indicating
+  a free page), skipping index 0 due to potential issues.
 
-  Attributes:
-    num_pages: Total number of available pages in the system
-    tokens_per_page: Number of tokens that can be stored in each page
-    slots: Number of sequence slots available for parallel processing
-    max_target_length: Maximum length of target sequences
-    max_prefill_predict_length: Maximum length for prefill prediction
-    max_pages_per_slot: Maximum number of pages that can be assigned to a slot
+  Args:
+    page_status: A 1D `jnp.ndarray` representing the global status of pages
+      (0 for free, 1 for allocated). Should have shape [num_pages].
+
+  Returns:
+    A scalar `jnp.int32` array containing the index of the next free page
+    (the lowest index >= 1 where `page_status` is 0).
+    Returns -1 if no free pages (at index >= 1) are found.
+  """
+  search_status = page_status[1:]
```

Comment:
Same here. Add a TODO gives some context on what's the issue with page 0, so we don't forget to resolve

Reply:
Will add.